cmd/codegen/generator/go/templates/module_types.go

```diff
@@ -0,0 +1,134 @@
+package templates
+
+import (
+	"fmt"
+	"go/types"
+
+	. "github.com/dave/jennifer/jen" // nolint:revive,stylecheck
+)
+
+// A Go type that has been parsed and can be registered with the dagger API
+type ParsedType interface {
+	// Generated code for registering the type with the dagger API
+	TypeDefCode() (*Statement, error)
+
+	// Go type referred to by this type
+	GoSubTypes() []types.Type
+}
+
+// parseGoTypeReference parses a Go type and returns a TypeSpec for the type *reference* only.
+// So if the type is a struct or interface, the returned TypeSpec will not have all the fields,
+// only the type name and kind.
+func (ps *parseState) parseGoTypeReference(typ types.Type, named *types.Named) (ParsedType, error) {
```

Comment:
nit: should we add a comment in here about why this doesn't create a `parseObjectType`? (it's since it's so that the caller of this can choose which objects to parse, but I'm not sure it's necessarily clear without the context of why that is).

Reply:
:tada: yup, thanks!